File: src/main/scala/Result.scala
package toy

sealed trait Result[+A] {
  import Result._

  def map[B](fn: A => B): Result[B] = {
    this match {
      case Success(s) => Success(fn(s))
      case Failure(f) => Failure(f)
    }
  }

  def flatMap[B](fn: A => Result[B]): Result[B] = {
    this match {
      case Success(s) => fn(s)
      case Failure(f) => Failure(f)
    }
  }
}

object Result {
  case class Success[A](value: A) extends Result[A]
  case class Failure(reason: String) extends Result[Nothing]
}

File: lingua-007/SwiftML/result.swift
//
//  SwiftML
//

indirect enum Result<S, F> {
  case Success(S)
  case Failure(F)

  public static func fromOptional(_ s: S?, _ f: F) -> Result<S, F> {
    switch s {
      case .some(let s): return .Success(s)
      case .none: return .Failure(f)
    }
  }

  func flatMap<T>(_ fn: (S) -> Result<T, F>) -> Result<T, F> {
    switch self {
      case .Success(let s): return fn(s)
      case .Failure(let f): return .Failure(f)
    }
  }
}
